### PR TITLE
Splits the rdp stream into read and write components, splits the loop up into read and write loops

### PR DESCRIPTION
**Pull request by ibeckermayer**

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 1]
```diff
@@ -132,101 +139,125 @@ impl Client {
...
+                    write_loop_handle.abort();
+                    Err(read_loop_panic.into())
+                }
+                futures_util::future::Either::Right((write_loop_panic, read_loop_handle)) => {
+                    read_loop_handle.abort();
+                    Err(write_loop_panic.into())
+                }
+            },
+        }
     }
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:202]: I think this should do it
```rust
tokio::select! {
    res = &mut read_loop_handle => {
        write_loop_handle.abort();
        res?
    },
    res = &mut write_loop_handle => {
        read_loop_handle.abort();
        res?
    }
}
```

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 2]
```diff
@@ -132,101 +139,125 @@ impl Client {
...
-                                self.handle_response_pdu(args).await?;
-                            },
-                            ClientFunction::Stop => {
-                                return Ok(());
-                            }
-                        }
+                let (action, mut frame) = read_stream.read_pdu().await?;
+                match action {
+                    ironrdp_pdu::Action::FastPath => unsafe {
+                        handle_remote_fx_frame(cgo_handle, frame.as_mut_ptr(), frame.len() as u32);
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:215]: We still block on `handle_remote_fx_frame` without telling tokio about it, right?

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 3]
```diff
@@ -104,8 +108,10 @@ impl Client {
 
         Ok(Self {
             cgo_handle,
-            rdp_stream,
-            x224_processor,
+            read_stream: Some(read_stream),
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/client.rs:112]: This is just a pattern I came up with wherein we can build up all the parts of the `Client`, then `take` ownership of all of them later to pass them into their respective read/write loops. Open to better ideas:

https://github.com/gravitational/teleport/blob/dab349960fd05ec5f0dea0f41a094cad7f25dfcf/lib/srv/desktop/rdp/rdpclient/src/client.rs#L144-L167

File: lib/srv/desktop/rdp/rdpclient/src/client.rs (hunk 1 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/client.rs:203]: Wow, I spent a lot of time trying to get something like that to no avail. Thanks 👍

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 4]
```diff
@@ -117,8 +123,9 @@ impl Client {
     /// and assigning the [`FunctionReceiver`] to `self.function_receiver`.
     fn register(mut self) -> Self {
         let (client_handle, function_receiver) = channel(100);
-        global::CLIENT_HANDLES.insert(self.cgo_handle, client_handle);
+        self.write_requester = Some(client_handle.clone());
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/srv/desktop/rdp/rdpclient/src/client.rs:127]: After second though let's leave it as is

File: lib/srv/desktop/rdp/rdpclient/Cargo.toml [hunk 5]
```diff
@@ -34,19 +34,19 @@ tokio-rustls =  { version = "0.24", features = ["dangerous_configuration"] }
 x509-parser = "0.14"
 sspi = { version = "0.10.1", features = ["network_client", "dns_resolver"] }
 
-ironrdp-connector = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
-ironrdp-tls = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
-ironrdp-session = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
-ironrdp-pdu = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
-ironrdp-tokio = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
+# ironrdp-connector = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
+# ironrdp-tls = { git = "https://github.com/Devolutions/IronRDP", rev = "babbd68af576f87044d99b2c091d3e0cf934d835" }
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/Cargo.toml:38]: Everything in this PR should work with regular IronRDP, right?

**Review by probakowski — APPROVED** [review]
LGTM, thanks! One small nit

**Pull request merged**